hazelcast/src/main/java/com/hazelcast/internal/dynamicclassloading/impl/ClassLocator.java

```diff
@@ -0,0 +1,202 @@
+import com.hazelcast.core.Member;
+import com.hazelcast.internal.cluster.ClusterService;
+import com.hazelcast.internal.dynamicclassloading.DynamicClassloadingService;
+import com.hazelcast.internal.dynamicclassloading.impl.filter.Filter;
+import com.hazelcast.internal.dynamicclassloading.impl.operation.ClassDataFinderOperation;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.util.ContextMutexFactory;
+
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+/**
+ * Provides classes to a local member.
+ *
+ * It's called by {@link com.hazelcast.internal.dynamicclassloading.DynamicClassLoader} when a class
+ * is not found on a local classpath.
+ *
+ * The current implementation can consult a cache and when the class is not found then it consults
+ * remote members.
+ *
+ */
+public final class ClassLocator {
+    private final ConcurrentMap<String, ClassSource> classSourceMap;
+    private final ClassLoader parent;
+    private final Filter<String> classNameFilter;
+    private final Filter<Member> memberFilter;
+    private final DynamicClassloadingConfig.ClassCacheMode classCacheMode;
+    private final NodeEngine nodeEngine;
+    private final ContextMutexFactory mutexFactory = new ContextMutexFactory();
+    private final ILogger logger;
+
+    public ClassLocator(ConcurrentMap<String, ClassSource> classSourceMap,
+                        ClassLoader parent, Filter<String> classNameFilter, Filter<Member> memberFilter,
+                        DynamicClassloadingConfig.ClassCacheMode classCacheMode, NodeEngine nodeEngine) {
+        this.classSourceMap = classSourceMap;
+        this.parent = parent;
+        this.classNameFilter = classNameFilter;
+        this.memberFilter = memberFilter;
+        this.classCacheMode = classCacheMode;
+        this.nodeEngine = nodeEngine;
+        this.logger = nodeEngine.getLogger(ClassLocator.class);
+    }
+
+    public static void onStartDeserialization() {
+        ThreadLocalClassCache.onStartDeserialization();
+    }
+
+    public static void onFinishDeserialization() {
+        ThreadLocalClassCache.onFinishDeserialization();
+    }
+
+    public Class<?> handleClassNotFoundException(String name)
+            throws ClassNotFoundException {
+        if (!classNameFilter.accept(name)) {
+            throw new ClassNotFoundException("Class " + name + " is not allowed to be loaded from other members.");
+        }
+        Class<?> clazz = tryToGetClassFromLocalCache(name);
+        if (clazz != null) {
+            return clazz;
+        }
+        return tryToGetClassFromRemote(name);
+    }
+
+    private Class<?> tryToGetClassFromRemote(String name) throws ClassNotFoundException {
+        ContextMutexFactory.Mutex classMutex = mutexFactory.mutexFor(name);
+        try {
+            synchronized (classMutex) {
```

Comment:
I am guessing that this is needed because you are not accessing it through the partition threads, can you document that here just to help the next one touching this code